File: solution/623. Add One Row to Tree/623_02.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int v, int d) {
        return dfs(root, v, d, true);
    }
    
private:
    TreeNode* dfs(TreeNode* node, int v, int d, bool isLeft) {
        if (d == 1) {
            auto new_node = new TreeNode(v);
            (isLeft ? new_node->left : new_node->right) = node;
            return new_node;
        }
        if (!node) return node;
        
        node->left = dfs(node->left, v, d - 1, true);
        node->right = dfs(node->right, v, d - 1, false);
        
        return node;
    }
};

File: solution/623. Add One Row to Tree/623_02.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        def dfs(node, d, is_left):
            if d == 1:
                new_node = TreeNode(val)
                if is_left:
                    new_node.left = node
                else:
                    new_node.right = node
                return new_node
            
            if not node:
                return node
            
            node.left = dfs(node.left, d - 1, True)
            node.right = dfs(node.right, d - 1, False)
            return node
        
        return dfs(root, depth, True)
    
